daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/processors/parsers/LayeredSequenceParser.scala

```diff
@@ -17,56 +17,48 @@
 
 package org.apache.daffodil.runtime1.processors.parsers
 
-import org.apache.daffodil.runtime1.layers.LayerExecutionException
-import org.apache.daffodil.runtime1.layers.LayerNotEnoughDataException
-import org.apache.daffodil.runtime1.layers.LayerRuntimeInfo
-import org.apache.daffodil.runtime1.layers.LayerTransformerFactory
+import org.apache.daffodil.runtime1.layers.LayerFactory
+import org.apache.daffodil.runtime1.layers.LayerRuntimeImpl
+import org.apache.daffodil.runtime1.layers.api.LayerUnexpectedException
 import org.apache.daffodil.runtime1.processors.SequenceRuntimeData
 
 class LayeredSequenceParser(
   rd: SequenceRuntimeData,
-  layerTransformerFactory: LayerTransformerFactory,
-  layerRuntimeInfo: LayerRuntimeInfo,
+  layerFactory: LayerFactory,
   bodyParser: SequenceChildParser,
 ) extends OrderedUnseparatedSequenceParser(rd, Vector(bodyParser)) {
   override def nom = "LayeredSequence"
 
-  override lazy val runtimeDependencies =
-    layerRuntimeInfo.evaluatables.toVector
-
   override def parse(state: PState): Unit = {
 
-    val layerTransformer = layerTransformerFactory.newInstance(layerRuntimeInfo)
+    // TODO: Separate the creation of layer into layerParse and layer
+    //   unparse. Right now they're blended onto one object.
+    //   It should be possible to define only a layer parser if a schema (and its required layers)
+    //   are intended to be used only to parse data.
+
+    val layerRuntimeImpl = new LayerRuntimeImpl(state, rd.layerRuntimeData)
+    val layerDriver = layerFactory.newInstance(layerRuntimeImpl)
     val savedDIS = state.dataInputStream
 
-    val isAligned = savedDIS.align(layerTransformer.mandatoryLayerAlignmentInBits, state)
+    val isAligned = savedDIS.align(layerDriver.mandatoryLayerAlignmentInBits, state)
     if (!isAligned)
       PE(
         state,
         "Unable to align to the mandatory layer alignment of %s(bits)",
-        layerTransformer.mandatoryLayerAlignmentInBits,
+        layerDriver.mandatoryLayerAlignmentInBits,
       )
 
     try {
-      val newDIS = layerTransformer.addLayer(savedDIS, state)
+      val newDIS = layerDriver.addInputLayer(savedDIS, layerRuntimeImpl)
 
       state.dataInputStream = newDIS
-      layerTransformer.startLayerForParse(state)
       super.parse(state)
-      layerTransformer.removeLayer(newDIS)
+      layerDriver.removeInputLayer(newDIS, layerRuntimeImpl)
     } catch {
-      case le: LayerNotEnoughDataException =>
-        PENotEnoughBits(
-          state,
-          le.schemaFileLocation,
-          le.dataLocation,
-          le.nBytesRequired * 8,
-        )
+      case pe: ParseError =>
+        state.setFailed(pe)
       case e: Exception =>
```

Comment:
Could a runtime SDE be caught here? Should that be treated differently than LayerUnexpectedException?